### ISSUE TITLE
Can BERT run on RTX 2070 * 2

### ISSUE THREAD

**Issue by guoyaohua** [google-research/bert#249]
In doc, the author suggest bert-base can run on RAM > 12 G GPU like 1080 or titan... But can BERT run on two RTX2070 which RAM is just 8G?  Did any friend had tried this? OR anybody can have other suggest ?

**Comment by guoyaohua** [comment]
THX for your help , but ..emm... the question i asked is not the distributed training, I want to know , can multi GPUs RAM can work for one model?  bert-base need >12G RAM GPU, but if i have two 8G GPUs, can it run on just one GPU, but use total 16G RAM of both GPUs?  not compute in both GPUs, just use them RAM, can it do this ?

**Comment by yyht** [comment]
you can apply with tf.device for different layers to achieve model parallsim with data parallism using horovod. i had implemented model parallism for en large base model on 4-gpus

**Comment by guoyaohua** [comment]
> you can apply with tf.device for different layers to achieve model parallsim with data parallism using horovod. i had implemented model parallism for en large base model on 4-gpus

Thanks for your reply.  Did you modify the core code of BERT? Make layers distribute in different GPUs , and can you success to load the pre-train data? 
You run BERT-Big model in 4-gpus? How many RAMs per GPU have?

**Comment by hanxiao** [comment]
for serving a pretrained/fine-tuned BERT model for feature extraction, you may try [bert-as-service](https://github.com/hanxiao/bert-as-service). It is optimized for efficiency, scalability as well as small memory footprint. The GPU memory consumption is around 700MB (idle) ~ 1.2GB (serving).

**Comment by guoyaohua** [comment]
> for serving a pretrained/fine-tuned BERT model for feature extraction, you may try [bert-as-service](https://github.com/hanxiao/bert-as-service). It is optimized for efficiency, scalability as well as small memory footprint. The GPU memory consumption is around 700MB (idle) ~ 1.2GB (serving).

Thanks ,
Dr. han, did you attempt to convert BERT from fp32 to fp16? It can save a half of memory, force cast the raw pre-trained params to fp16 will punish model quality？

**Comment by hanxiao** [comment]
actually, i’m working on more aggressive quantization using 4bit integer. float16 is of course also interesting but it’s not my priority now.

stay tune on bert-as-service, feel free to contribute if you achieve something with float16.

**Comment by hanxiao** [comment]
@guoyaohua FYI, since 1.7.4 `bert-as-service` supports FP16. However, it highly depends on your GPU and whether it supports FP16 instructions. On my Tesla M40 with no FP16 support, the inference speed is 20x times (`bert-serving-start -fp16`) slower than float32.

For the install/update, please do `pip install -U bert-serving-server bert-serving-client`.

**Comment by guoyaohua** [comment]
> @guoyaohua FYI, since 1.7.4 `bert-as-service` supports FP16. However, it highly depends on your GPU and whether it supports FP16 instructions. On my Tesla M40 with no FP16 support, the inference speed is 20x (`bert-serving-start -fp16`) **slower** than float32.
> 
> Anyway, a feature is a feature, could be useful to others. For the install/update, please do `pip install -U bert-serving-server bert-serving-client`.

Thank you, Does this impair the performance of the model?

**Comment by hanxiao** [comment]
No, `fp16` is just a CLI option, you can turn on/off as you wish. I always argue `bert-as-service` is the inference library that you want to use in production, with 900 sentences/s. See https://github.com/hanxiao/bert-as-service#zap-benchmark

**Comment by bcbcbcbcbcl** [comment]
With one rtx 2070, only able to train with batch_size = 8, max_seq_length = 320.

**Comment by zheolong** [comment]
> you can apply with tf.device for different layers to achieve model parallsim with data parallism using horovod. i had implemented model parallism for en large base model on 4-gpus

Can u share ur code please, that will be helpfull, thx

**Comment by disign-ming** [comment]
How does 1080 8g fine-turning bert small locally?

**Comment by bbreton3** [comment]
> How does 1080 8g fine-turning bert small locally?
batch_size = 8, max_seq_length = 320, does a full fine-tuning on french wikipedia in 6 hours.
